### PR TITLE
Root nodes for the duration of their CSS transitions

### LINKED ISSUE #14972: TransitionEnd event can include nodes that have been GCed
https://github.com/servo/servo/commits/5d3847dddc9bb7907abfa5d38a7927d6c656fbc1/components/layout/animation.rs sends an asynchronous message to the script thread that contains an UnsafeNode value. When the script thread processes the message, it casts the unsafe node to a DOM node and tries to use it. If the node has been GCed since the layout event that caused the message to be sent, this is a crash.

I propose fixing this in the same way that #14962 deals with asynchronous operations involving DOM nodes - collect them during layout and have the script thread retrieve them immediately to prevent risk of them being GCed.

I hit this in a debugger while running the TodoMVC benchmark: https://lhorie.github.io/todomvc-perf-comparison/todomvc-benchmark/

### PR DESCRIPTION
**Pull request by jdm**
This ensures that we can pass a node address as part of the asynchronous
transition end notification, making it safe to fire the corresponding
DOM event on the node from the script thread. Without explicitly rooting
this node when the transition starts, we risk the node being GCed before
the transition is complete.

---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #14972
- [X] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16295)

### CODE REVIEW

**Comment by jdm** [comment]
As I feared, there could be problems with this approach when the animations are triggered from the compositor:
```
  ▶ CRASH [expected TIMEOUT] /dom/events/EventListener-invoke-legacy.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ assertion failed: rw_data.newly_transitioning_nodes.is_empty() (thread LayoutThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }, at /Users/servo/buildbot/slave/mac-rel-wpt1/build/components/layout_thread/lib.rs:1411)
  │ 2017-04-07 01:35:16.662 servo[70066:646481199] Metadata.framework [Error]: couldn't get the client port
  │ stack backtrace:
  │    0:        0x10459e49e - backtrace::backtrace::trace::h47a32bf8c6f00c9b
  │    1:        0x10459ebec - backtrace::capture::Backtrace::new::hb5a725a088a2a2fc
  │    2:        0x102e7d5b8 - servo::main::_$u7b$$u7b$closure$u7d$$u7d$::hab3ba3e402ac1f0b
  │    3:        0x1049ff9c9 - std::panicking::rust_panic_with_hook::h05996066754c6be9
  │    4:        0x102faa4f4 - std::panicking::begin_panic::haab64c72e0aa1aab
  │    5:        0x10300ecdc - layout_thread::LayoutThread::perform_post_style_recalc_layout_passes::h139d5d3cd94b81e4
  │    6:        0x10300d2d5 - layout_thread::LayoutThread::tick_all_animations::h90b400a44eeb0e6f
  │    7:        0x103003306 - layout_thread::LayoutThread::handle_request_helper::h307a64d67c3c4130
  │    8:        0x102f944a0 - profile_traits::mem::ProfilerChan::run_with_memory_reporting::h6184f9f4fcb8e847
  │    9:        0x103001173 - _$LT$layout_thread..LayoutThread$u20$as$u20$layout_traits..LayoutThreadFactory$GT$::create::_$u7b$$u7b$closure$u7d$$u7d$::h8028856a97c8c244
  │   10:        0x102faadfd - std::panicking::try::do_call::ha664f0ed882ba9a8
  │   11:        0x104a0099a - __rust_maybe_catch_panic
  │   12:        0x102fc41e6 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::hc02c54adca1db771
  │   13:        0x1049fc4e4 - std::sys::imp::thread::Thread::new::thread_start::h4008e1859fbd98b8
  │   14:     0x7fff8af9f059 - _pthread_body
  │   15:     0x7fff8af9efd6 - _pthread_start
  │ ERROR:servo: assertion failed: rw_data.newly_transitioning_nodes.is_empty()
  │ Pipeline failed in hard-fail mode.  Crashing!
  └ thread panicked while processing panic. aborting.
```

**Comment by jdm** [comment]
@nox do you want to review this?

File: components/script/dom/window.rs [hunk 1]
```diff
@@ -1244,6 +1245,9 @@ impl Window {
             }
         }
 
+        let newly_transitioning_nodes = self.layout_rpc.newly_transitioning_nodes();
```
Comment (nox) [inline components/script/dom/window.rs]: Layout can send back garbage addresses and then everything goes to hell.

File: components/script/dom/window.rs [hunk 2]
```diff
@@ -1244,6 +1245,9 @@ impl Window {
...
+        ScriptThread::note_newly_transitioning_nodes(newly_transitioning_nodes);
```
Comment (nox) [inline components/script/dom/window.rs]: This function definitely should be unsafe.

File: components/script/script_thread.rs [hunk 3]
```diff
@@ -574,6 +576,17 @@ impl ScriptThreadFactory for ScriptThread {
 }
 
 impl ScriptThread {
+    pub fn note_newly_transitioning_nodes(nodes: Vec<UntrustedNodeAddress>) {
+        SCRIPT_THREAD_ROOT.with(|root| {
+            let script_thread = unsafe { &*root.get().unwrap() };
+            let js_runtime = script_thread.js_runtime.rt();
+            let new_nodes = nodes
+                .into_iter()
+                .map(|n| JS::from_ref(&*from_untrusted_node_address(js_runtime, n)));
```
Comment (nox) [inline components/script/script_thread.rs]: Function `from_untrusted_node_address` should have definitely be `unsafe`.

File: components/script/script_thread.rs [hunk 4]
```diff
@@ -1600,11 +1615,27 @@ impl ScriptThread {
     }
 
     /// Handles firing of transition events.
-    #[allow(unsafe_code)]
-    fn handle_transition_event(&self, unsafe_node: UnsafeNode, name: String, duration: f64) {
-        let node = unsafe { ServoLayoutNode::from_unsafe(&unsafe_node) };
-        let node = unsafe { node.get_jsmanaged().get_for_script() };
-        let window = window_from_node(node);
+    fn handle_transition_event(&self, unsafe_node: UntrustedNodeAddress, name: String, duration: f64) {
```
Comment (nox) [inline components/script/script_thread.rs]: This should be unsafe.

File: components/layout/animation.rs [hunk 5]
```diff
@@ -86,8 +89,8 @@ pub fn update_animation_state(constellation_chan: &IpcSender<ConstellationMsg>,
                 continue
             }
 
-            if let Animation::Transition(_, unsafe_node, _, ref frame, _) = running_animation {
-                script_chan.send(ConstellationControlMsg::TransitionEnd(unsafe_node,
+            if let Animation::Transition(node, _, ref frame, _) = running_animation {
+                script_chan.send(ConstellationControlMsg::TransitionEnd(node.to_untrusted_node_address(),
```
Comment (nox) [inline components/layout/animation.rs]: Why can't we send back the `node` itself?

File: components/script/script_thread.rs [hunk 6]
```diff
@@ -574,6 +576,17 @@ impl ScriptThreadFactory for ScriptThread {
 }
 
 impl ScriptThread {
+    pub fn note_newly_transitioning_nodes(nodes: Vec<UntrustedNodeAddress>) {
```
Comment (nox) [inline components/script/script_thread.rs]: Can GC happen before this even gets a change to be called?

IMO this whole function should be unsafe.

**Review by nox — CHANGES_REQUESTED** [review]
This does not look as safe as I would like to be but this is certainly better than the code code we currently have.

That being said, I don't think we will ever have sound transition code for as long as we continue to send `UnsafeNodeAddress`-like types between layout and script.

File: components/script/script_thread.rs (hunk 6 above)
Comment (jdm) [inline components/script/script_thread.rs]: Anything that has a transition starting during layout is part of the DOM, therefore cannot be GCed. There's no need for this to be unsafe.

File: components/layout/animation.rs (hunk 5 above)
Comment (jdm) [inline components/layout/animation.rs]: It's a layout type.

File: components/script/script_thread.rs (hunk 6 above)
Comment (nox) [inline components/script/script_thread.rs]: It needs to be unsafe.

```thread
ScriptThread::note_newly_transitioning_nodes(vec![UntrustedNodeAddress:from_id(0xdeadbeaf))
```

**Comment by jdm** [comment]
Addressed comments about methods that should be unsafe.

File: components/script/dom/node.rs [hunk 7]
```diff
@@ -927,20 +927,18 @@ fn first_node_not_in<I>(mut nodes: I, not_in: &[NodeOrString]) -> Option<Root<No
...
 /// returns it.
 #[allow(unsafe_code)]
-pub fn from_untrusted_node_address(_runtime: *mut JSRuntime, candidate: UntrustedNodeAddress)
+pub unsafe fn from_untrusted_node_address(_runtime: *mut JSRuntime, candidate: UntrustedNodeAddress)
     -> Root<Node> {
-    unsafe {
-        // https://github.com/servo/servo/issues/6383
-        let candidate: uintptr_t = mem::transmute(candidate.0);
+    // https://github.com/servo/servo/issues/6383
+    let candidate: uintptr_t = mem::transmute(candidate.0);
```
Comment (nox) [inline components/script/dom/node.rs]: Please don't use transmute.

File: components/script/dom/node.rs [hunk 8]
```diff
@@ -927,20 +927,18 @@ fn first_node_not_in<I>(mut nodes: I, not_in: &[NodeOrString]) -> Option<Root<No
...
 //        let object: *mut JSObject = jsfriendapi::bindgen::JS_GetAddressableObject(runtime,
 //                                                                                  candidate);
-        let object: *mut JSObject = mem::transmute(candidate);
-        if object.is_null() {
-            panic!("Attempted to create a `JS<Node>` from an invalid pointer!")
-        }
-        let boxed_node = conversions::private_from_object(object) as *const Node;
-        Root::from_ref(&*boxed_node)
+    let object: *mut JSObject = mem::transmute(candidate);
```
Comment (nox) [inline components/script/dom/node.rs]: Please don't use transmute.

**Review by nox — APPROVED** [review]
Looks good, just two nits, feel free to not address them if you don't care.

**Pull request merged**